src/vos/vos_aggregate.c

```diff
@@ -688,12 +637,54 @@ fill_one_segment(daos_handle_t ih, struct agg_merge_window *mw,
 		mark_yield(&addr_src, acts);
 		D_ASSERT(biov_idx < bsgl.bs_nr);
 		bio_iov_set(&bsgl.bs_iovs[biov_idx], addr_src, copy_size);
-		biov_idx++;
 
+		if (mw->mw_csum_support) {
+			unsigned int wider = 0; /* length of per-ext csum add */
+			unsigned int add_cnt =
+				vos_csum_widen_biov(&bsgl.bs_iovs[biov_idx],
+						    phy_ent, &ext,
+						    ent_in->ei_inob, phy_lo,
+						    &wider);
+
+			/* add_cnt is the number of data segments to add. */
+			if (add_cnt) {
+				buf_add += wider;
+				iov.iov_buf_len += wider;
+				copy_size += wider;
+				added_csum_segs += add_cnt;
+			}
+			vos_csum_add_recalcs(&csum_recalcs, phy_ent, &ext,
+					     &bsgl, biov_idx);
+		}
+		biov_idx++;
 		D_ASSERT(iov.iov_buf_len >= copy_size);
 		iov.iov_buf_len -= copy_size;
 	}
-	D_ASSERT(seg_size == (io->ic_buf_len - iov.iov_buf_len));
+
+	D_ASSERT(seg_size == buf_max - iov.iov_buf_len);
+
+	/* Moved read buf allocation to after loop, to allow inclusion
+	 * of additional data needed for verification of prior checksums.
+	 */
```

Comment:
Isn't the total size being calculated in prepare phase? (prepare_segments())

Reply:
The size of the buffer for the readv has to be extended to accommodate additional data required for checksum verification.  The extended sizing is calculated as the read extents are formulated, and this quantity is not know when the segments are prepared.